torch/_dynamo/utils.py

```diff
@@ -3106,3 +3107,43 @@ def get_user_object_from_id(obj_id):
 def store_user_object_weakref(obj):
     obj_id = id(obj)
     user_obj_id_to_weakref[obj_id] = weakref.ref(obj)
+
+
+class CompileTimeInstructionCounter:
+    _counter: int = 0
+    _id: int = -1
+
+    @classmethod
+    def start(cls) -> None:
+        if cls._id != -1:
+            raise RuntimeError("CompileTimeInstructionCounter is already started")
+        cls._id = i_counter.start()
+
+    @classmethod
+    def end(cls) -> None:
+        cls._counter += i_counter.end(cls._id)
+        cls._id = -1
+
+    @classmethod
+    def clear(cls) -> None:
+        cls._counter = 0
+
+    @classmethod
+    def value(cls) -> int:
+        return cls._counter
+
+
+class CompileTimeInstructionCollector:
+    def __init__(self):
+        pass
+
+    def __enter__(self):
+        if config.record_compile_time_instruction_count:
+            self.started = True
```

Comment:
Do we want this to be able to handle recursive calls?
```
with CompileTimeInstructionCollector():
  ...
  with CompileTimeInstructionCollector():
    ...
```
Is so might want a counter instead of a bool...

Reply:
right now its not used with recursions but i can make it resilinet to that